### ISSUE TITLE
[Question] Is it possible to extend the existing metadata directory to add more metrics to it?

### ISSUE THREAD

**Issue by raphaelTrench** [automl/auto-sklearn#1225]
# Short Question Description
Is it possible to extend the existing metadata directory to add more metrics to it?

Hello, first of all congratulations on the project!
I am having a bit of  trouble understanding how to properly  use the `metadata_directory` parameter. Mainly for two reasons:  
1) When I try to input a search metric autosklearn has no metadata information of, I get the error(in this case binary_ks):
`[WARNING] [2021-08-19 17:55:34,151:Client-AutoMLSMBO(42)::cbc6aea0-012f-11ec-ae49-367ddad02bc4] Could not find meta-data directory /Users/c91195a/Library/Caches/pypoetry/virtualenvs/dragon-oQHUJD0o-py3.8/lib/python3.8/site-packages/autosklearn/metalearning/files/binary_ks_stat_binary.classification_dense`  
And that would make sense for the first time, but even after the experiment is finished, there is no new new `binary_ks_stat_binary.classification_dense` directory.

2) When inputting the path to an empty directory to the `metadata_directory` parameter, the autosklearn fails as it was already expecting the metadata files:
```python
ERROR] [2021-08-19 18:11:35,480:autosklearn.metalearning.input.aslib_simple] Not found: ml/metadata/binary_ks_stat_binary.classification_dense/algorithm_runs.arff (has to be added)
[ERROR] [2021-08-19 18:11:35,480:autosklearn.metalearning.input.aslib_simple] Not found: ml/metadata/binary_ks_stat_binary.classification_dense/feature_values.arff (has to be added)
Traceback (most recent call last):
  File "/Users/c91195a/Library/Caches/pypoetry/virtualenvs/dragon-oQHUJD0o-py3.8/bin/dragon", line 5, in <module>
[ERROR] [2021-08-19 18:11:35,480:Client-AutoML(42):08c7b57c-0132-11ec-9a09-367ddad02bc4] 'NoneType' object is not iterable
Traceback (most recent call last):
  File "/Users/c91195a/Library/Caches/pypoetry/virtualenvs/dragon-oQHUJD0o-py3.8/lib/python3.8/site-packages/autosklearn/automl.py", line 767, in fit
    _proc_smac.run_smbo()
  File "/Users/c91195a/Library/Caches/pypoetry/virtualenvs/dragon-oQHUJD0o-py3.8/lib/python3.8/site-packages/autosklearn/smbo.py", line 406, in run_smbo
    metalearning_configurations = self.get_metalearning_suggestions()
  File "/Users/c91195a/Library/Caches/pypoetry/virtualenvs/dragon-oQHUJD0o-py3.8/lib/python3.8/site-packages/autosklearn/smbo.py", line 577, in get_metalearning_suggestions
    meta_base = MetaBase(self.config_space, self.metadata_directory, self.logger)
  File "/Users/c91195a/Library/Caches/pypoetry/virtualenvs/dragon-oQHUJD0o-py3.8/lib/python3.8/site-packages/autosklearn/metalearning/metalearning/meta_base.py", line 46, in __init__
    for algorithm_id in self.configurations:
TypeError: 'NoneType' object is not iterable
```

So in short, Id liek to understand better how to generate the needed file structure and how to update it as I run experiments.
thank you in advance!

# System Details (if relevant)
* Which version of `auto-sklearn` are you using?
0.13.0
* Are you running this on Linux / Mac / ... ?
Mac

**Comment by mfeurer** [comment]
Hey, thanks a lot for your interest in Auto-sklearn. Auto-sklearn does not update the meta-data over time, which is why you observe 2. As we ship pre-defined meta-data, you observe 1., because we cannot ship every metric.

If you want to learn more about how the data for Auto-sklearn 1.0 is generated you could start by checking the accompanying unit test: https://github.com/automl/auto-sklearn/blob/master/test/test_scripts/test_metadata_generation.py

Regarding Auto-sklearn 2.0 - it is definitely possible, but much harder and you can have a look at the scripts we provide at https://github.com/automl/ASKL2.0_experiments

**Comment by raphaelTrench** [comment]
Thanks for the info!
Would it be a useful contribution to the project to add meta-data from other common metrics? Following the same process presented in the `script` directory of course.
The same question also goes for auto-sklearn 2.0 policies per-metric.

**Comment by mfeurer** [comment]
What metrics would you have in mind? At least for Auto-sklearn 1.0 it would be rather simple to add them so they would be available in the next re-calculation.

For Auto-sklearn 2.0 it's a bit tougher because that requires quite some compute and we don't really have an automated process for that yet.

**Comment by raphaelTrench** [comment]
The  metrics of most interest to me at the moment are the binary ks and gini entropy. They are the main metrics to be  optimized in most projects where I work, and I believe in general in a lot credit problems (at least in my region).

**Comment by mfeurer** [comment]
Are these metrics available in scikit-learn? If yes, we can add them to the list of metrics we generate meta-data for.

**Comment by raphaelTrench** [comment]
They are not avaiable in sklearnis thata blocker? But [ks is avaiable as a function in scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ks_2samp.html)


And gini could be calculated as 
```python
2 * sklearn.metrics.roc_auc_score(y_true, y_pred) - 1
```

We have manually been using them with autosklearn using its `make_scorer()` function.
If possible, I also believe I could make a contribution to add them

**Comment by mfeurer** [comment]
Since gini is a linear transformation of the ROC AUC you can use Auto-sklearn's ROC AUC metric and get similar results. In order to have the ks score it would be great if you could [add this to scikit-learn via a PR](https://github.com/scikit-learn/scikit-learn/discussions/20466) and we'll add it once there is a scikit-learn metric. In the meantime I propose the following hack for you: take one of the directories under `autosklearn/metalearning/files/`, copy it, and rename the metric to match your metric name. By that, auto-sklearn should pick up on it and improve results until we have a proper fix for this.
